builtin/providers/aws/resource_aws_security_group.go

```diff
@@ -675,36 +682,102 @@ func resourceAwsSecurityGroupUpdateRules(
 	return nil
 }
 
+func clearSecurityGroupCache(cache *SecurityGroupsCache) {
+	mtKey := "security_groups_cache"
+	awsMutexKV.Lock(mtKey)
+	defer awsMutexKV.Unlock(mtKey)
+	cache.SecurityGroups = nil
+}
+
+func updateSecurityGroupCache(conn *ec2.EC2, cache *SecurityGroupsCache) error {
+	mtKey := "security_groups_cache"
+	awsMutexKV.Lock(mtKey)
+	defer awsMutexKV.Unlock(mtKey)
+	if cache.SecurityGroups == nil {
+		req := &ec2.DescribeSecurityGroupsInput{}
+		resp, err := conn.DescribeSecurityGroups(req)
+
+		if err != nil {
+			log.Printf("Error on SGStateRefresh: %s", err)
+			return err
+		}
+
+		cache.SecurityGroups = resp
+	}
+	return nil
+}
+
 // SGStateRefreshFunc returns a resource.StateRefreshFunc that is used to watch
 // a security group.
-func SGStateRefreshFunc(conn *ec2.EC2, id string) resource.StateRefreshFunc {
+func SGStateRefreshFunc(conn *ec2.EC2, cache *SecurityGroupsCache, id string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
-		req := &ec2.DescribeSecurityGroupsInput{
-			GroupIds: []*string{aws.String(id)},
+		group, err := RefreshSG(conn, cache, id)
+		switch {
+		case group != nil:
+			return group, "exists", nil
+		case err != nil:
+			return nil, "", err
+		default:
+			return nil, "", nil
 		}
-		resp, err := conn.DescribeSecurityGroups(req)
-		if err != nil {
-			if ec2err, ok := err.(awserr.Error); ok {
-				if ec2err.Code() == "InvalidSecurityGroupID.NotFound" ||
-					ec2err.Code() == "InvalidGroup.NotFound" {
-					resp = nil
-					err = nil
-				}
-			}
+	}
+}
+
+func RefreshSG(conn *ec2.EC2, cache *SecurityGroupsCache, id string) (*ec2.SecurityGroup, error) {
```

Comment:
Can we make functions like `RefreshSG`, `RefreshSGFromCache`, `RefreshSGFromAPI` et. al. should be lowercase and unexpected?